lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    ch = {
+      contextPreviewExists: false
+
+      removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while editor.session.getContext(i - 1) == "equation"
+          i -= 1
+        start = i
+        while editor.session.getContext(i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
```

Comment:
So, this hack with context probes and filtering with regexp is okay for the proof of the concept; however in production we should probably use more reliable techniques: either our own or KaTeX's tokenizers. 

Also, it seems that KaTeX can just render unknown macros as text, maybe that's a good trade-off?